File: 437. Path Sum III/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) int {
	count := 0
	sumsMap := make(map[int]int)
	sumsMap[0] = 1
	traverse(&count, root, 0, sumsMap, sum)
	return count
}

func traverse(count *int, currNode *TreeNode, currSum int, sums map[int]int, sum int) {
	if currNode == nil {
		return
	}

	currSum += currNode.Val
	if c, okay := sums[currSum-sum]; okay {
		*count += c
	}

	if _, okay := sums[currSum]; okay {
		sums[currSum]++
	} else {
		sums[currSum] = 1
	}

	traverse(count, currNode.Left, currSum, sums, sum)
	traverse(count, currNode.Right, currSum, sums, sum)

	sums[currSum]--
}

File: 437. Path Sum III/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        def dfs(node, current_sum):
            self.count += path_freqs.get(current_sum - sum, 0)
            path_freqs[current_sum] = path_freqs.get(current_sum, 0) + 1
            if node.left:
                dfs(node.left, current_sum + node.left.val)
            if node.right:
                dfs(node.right, current_sum + node.right.val)
            path_freqs[current_sum] = path_freqs.get(current_sum, 0) - 1

        self.count = 0
        path_freqs = {0: 1}

        if root:
            dfs(root, root.val)
        return self.count
